### PR TITLE
Spark BulkWriter fix to ensure re-enqueueing retries

### PR DESCRIPTION
**Pull request by FabianMeiswinkel**
# Description

This PR fixes a race condition when retries were not re-enqueue for bulk writes in some situations. The PR also adds test coverage by adding a way to use FaultInjection from Spark.

Last-but-not-least this PR fixes a possible `NullPointerException` in the ctor of `FeedOperationState`

## `CosmosClientBuilderInterceptor` service trait
This new trait is added to allow dependency injection based interception of CosmosCLinetBuilder creation in the Spark connector. It allows for example to inject custom diagnostics handlers etc. - this trait is used in the tests to create assertions based on the observed diagnostics. When a new `CosmosClientBuilder` is created to create a new CosmosClient instance, the interceptors provided to the `spark.cosmos.account.clientBuilderInterceptors` config are applied.

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/BulkWriter.scala [hunk 1]
```diff
@@ -989,14 +994,25 @@ private class BulkWriter(container: CosmosAsyncContainer,
                   numberOfIntervalsWithIdenticalActiveOperationSnapshots
                 )
 
-                if (numberOfIntervalsWithIdenticalActiveOperationSnapshots.get == 1L &&
-                    // pending retries only tracks the time to enqueue
-                    // the retry - so this should never take longer than 1 minute
-                    pendingRetriesSnapshot == 0L &&
-                    Scannable.from(bulkInputEmitter).scan(Scannable.Attr.BUFFERED) >0) {
+                if (numberOfIntervalsWithIdenticalActiveOperationSnapshots.get > 0L) {
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/BulkWriter.scala:997]: just to confirm, so we also want to loose how many times we gonna requeue here?

**Review by xinlian12 — APPROVED** [review]
LGTM, thanks

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/BulkWriter.scala (hunk 1 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/BulkWriter.scala:997]: yes, I doubt it helps at that point - but won't hurt either.

**Pull request merged**